### ISSUE TITLE
Question about avoiding compiler's notes re expansion macros for cpp_magic.

### ISSUE THREAD

**Issue by vsolontsov-ll** [18sg/uSHET#6]
Hi,

First of all many thanks for your efforts!

I'm using cpp_magic to implement yet another (surely, super fast:-)) logger in C (not C++ where variadic templates help a lot) and the only issue I have is that any error in the use of my macro generates thousands of `note: in expansion of macro ‘EVAL...'` messages. 

I can reduce the number of those notes by adding own MAP-like macros based on EVAL64 -- that's the minimal in my case. But still it gives 400+ notes. So I'm wondering if you have a solution or advice for this issue?

NB I tried to find a gcc diagnostic option (either tweakable from pragmas or acceptable for the whole project) but I failed. Currently I'm thinking about adding some basic macro arithmetic (DEC_/INC_) to count macro arguments and use a limited 0-9 args version for 99% of cases while leave the full-up-to-64 for the reset.

**Comment by mossblaser** [comment]
Hello there! I'm glad to see you're having fun with the library!

I'm afraid I don't think there is a sensible solution to the extremely long error message situation except, as you've found, using smaller numbers of `EVAL` passes. This is ultimately the kind of price you have to pay when doing something as dirty as this :).

When using this library in EZSHET I found the only practical development process was to:

1) not make any mistakes, and...
2) if I did make a mistake, treat the compiler output as just an opaque "NOPE!" and solve the problem by just staring at what I'd written till I found the mistake.

In any case, I think the choice you have to make is really whether using this library is actually a good idea or you ought to try something else. If you're doing this purely for fun -- go for it! This library only exists because it was fun, I hope you have some too! But think very hard about including it in anything real: if you think you're finding debugging now tough, now imagine encountering this in the wild in a non-trivial complete system... (Of course, despite my repeating this advice everywhere, I've hard of the likes of Apple and even some proposed Linux kernel patches are using it...).

More concretely, it sounds like if you're contented with hard-coding macros to cope with cases of up to 9 arguments, you'd be just as well to handle (say) 0-64 arguments in the same way. Just write some shell/python/whatever to generate the lines you need and you're done. Its ugly but it is, at least, easily understood.

Hope that helps and I do hope I don't put you off having fun! :D. If you do come up with a neat trick (be it a lesser known compiler option or a neat sed/awk filter) I'd like to be first to know!

All the best,

Jonathan

**Comment by vsolontsov-ll** [comment]
(Re-posted with markdown)

**Comment by vsolontsov-ll** [comment]
Hey Jonathan,

Thanks for the prompt reply and for your advice. As you were willing to
know, let me share what I came to yesterday :-). It's not perfect but at
least for a small number of arguments it doesn't generate more than
meaninful expansion notes.
TLDR: https://godbolt.org/z/b45vMor3r

First of all, honestly I'm not so skeptical about the approach. I rolled
out some code on boost::mlp into prod back in 2013 (we had to use some gcc
4.8 in pre-C++11 mode) and as far as I know it's still in use and supported.

I started with hardcoding a macro accepting up to 64 (sure initially I had
just a few as a POC). That was a typical `my_macro(__VA_ARGS__
__VA_OPT__(,) fake_arg, fake_arg, /*... 64 times */)`.
There's an obvious issue: despite the majority use-cases being just a few
args an expansion of every such macro contains the code for all 64 -- it's
huge and messy partially because for every argument there's a large
_Generic(...) construction.
Now I realize that I was able to expand operations on the `fake_arg` to
nothing by using some tricks. Yet... using tricks which are not very
readable.

Your MAP-like stuff does the trick nicely and generates no fake C code
without extra tricks.

Now what I drafted yesterday...
1. Added basic counting 1...8, MANY:
```C++
  #define INC(x) CAT(INC_, x)
  #define INC_0 1
  #define INC_1 2
  #define INC_2 3
  #define INC_3 4
  #define INC_4 5
  #define INC_5 6
  #define INC_6 7
  #define INC_7 8
  #define INC_8 MANY
  #define INC_MANY MANY
```
2. Added basic check for MANY
```C++
#define NOT_MANY(x) IS_PROBE(CAT(_NOT_MANY_, x))
#define _NOT_MANY_MANY PROBE()
/**
* Check if x is MANY
*/
#define IS_MANY(x) NOT_MANY(x)
```
3. Added missing EVALn for n in {3, 5, 6, 7} and EVALMANY
```C++
#define EVAL7(...) EVAL3(EVAL4(__VA_ARGS__))
#define EVAL6(...) EVAL2(EVAL4(__VA_ARGS__))
#define EVAL5(...) EVAL1(EVAL4(__VA_ARGS__))
#define EVAL4(...) EVAL2(EVAL2(__VA_ARGS__)) /* Already existed */
#define EVAL3(...) EVAL1(EVAL2(__VA_ARGS__))
#define EVALMANY(...) EVAL(__VA_ARGS__)
```
4. Added a macro to count arguments from 0 to MANY
```C++
#define COUNT_ARGS(...) \
IF_ELSE(HAS_ARGS(__VA_ARGS__))(EVAL8(COUNT_ARGS_INNER(1, ##__VA_ARGS__)), 0)
#define COUNT_ARGS_INNER(n, first, ...) \
IF_ELSE(OR(NOT(HAS_ARGS(__VA_ARGS__)), IS_MANY(n)))(n, DEFER2
(_COUNT_ARGS_INNER)()(INC(n), ##__VA_ARGS__))
#define _COUNT_ARGS_INNER() COUNT_ARGS_INNER
```
5. Added trivial CA_INDIR for "indirectly" concatenate
```C++
#define CAT_INDIR(a, ...) CAT(a, __VA_ARGS__)
```
6. Finally modified your MAP macro to use EVALn:
```C++
#define MAP(op,sep,...) \
IF(HAS_ARGS(__VA_ARGS__))(CAT_INDIR(EVAL, COUNT_ARGS(__VA_ARGS__))(MAP_INNER
(op,sep,__VA_ARGS__)))
#define MAP_INNER(op,sep,cur_val, ...) \
op(cur_val) \
IF(HAS_ARGS(__VA_ARGS__))( \
sep() DEFER2(_MAP_INNER)()(op, sep, ##__VA_ARGS__) \
)
#define _MAP_INNER() MAP_INNER
```
I just realized there's probably a bit better way, giving a user more
freedom: add MAP_LIMITED(limit, op, sep, ...).
Then the user can pass either a value of COUNT_ARGS or his worst case
limit. Also it won't be depending on `COUNT_ARGS` which can be implemented
for a different threshold value.

Cheers,
Vlad
